runtime/compiler/optimizer/J9ValuePropagation.cpp

```diff
@@ -1841,24 +2176,28 @@ J9::ValuePropagation::doDelayedTransformations()
             callTree->insertBefore(TR::TreeTop::create(comp(), compressNode));
             }
          }
-      else
+      else if (isStore && !isFlattenedElement)
          {
+         TR::Node *elementAddressNode = J9::TransformUtil::calculateElementAddress(comp(), arrayRefNode, indexNode, TR::Address);
          TR::Node *oldAnchorNode = callTree->getNode();
 
          TR_ASSERT_FATAL_WITH_NODE(oldAnchorNode, (oldAnchorNode->getNumChildren() == 1) && oldAnchorNode->getFirstChild() == callNode, "Expected call node n%un for jitStoreFlattenableArrayElement was anchored under node n%un\n", callNode->getGlobalIndex(), oldAnchorNode->getGlobalIndex());
 
          TR::Node *elementStoreNode = TR::Node::recreateWithoutProperties(callNode, TR::awrtbari, 3, elementAddressNode,
                                                    valueNode, arrayRefNode, elementSymRef);
 
-         if (needsStoreCheck || needsStoreAndNullCheck)
+         if (needsStoreCheck || needsNullValueCheck)
             {
-            TR::ResolvedMethodSymbol *methodSym = comp()->getMethodSymbol();
-            TR::SymbolReference *storeCheckSymRef = comp()->getSymRefTab()->findOrCreateTypeCheckArrayStoreSymbolRef(methodSym);
-            TR::Node *storeCheckNode = TR::Node::createWithRoomForThree(TR::ArrayStoreCHK, elementStoreNode, 0, storeCheckSymRef);
-            storeCheckNode->setByteCodeInfo(elementStoreNode->getByteCodeInfo());
-            callTree->setNode(storeCheckNode);
+            if (needsStoreCheck)
+               {
+               TR::ResolvedMethodSymbol *methodSym = comp()->getMethodSymbol();
+               TR::SymbolReference *storeCheckSymRef = comp()->getSymRefTab()->findOrCreateTypeCheckArrayStoreSymbolRef(methodSym);
+               TR::Node *storeCheckNode = TR::Node::createWithRoomForThree(TR::ArrayStoreCHK, elementStoreNode, 0, storeCheckSymRef);
+               storeCheckNode->setByteCodeInfo(elementStoreNode->getByteCodeInfo());
+               callTree->setNode(storeCheckNode);
```

Comment:
Previously, this call to `setNode` would replace the helper call if `needsStoreCheck || needsNullValueCheck` was true, and [the call to `setNode` in the `else` branch below](https://github.com/eclipse-openj9/openj9/blob/e81ca9bb8846805dfb6f65db572ba4baac307659/runtime/compiler/optimizer/J9ValuePropagation.cpp#L2213-L2216) would replace the helper call if `needsStoreCheck || needsNullValueCheck` was false.

With the current version, if `needsStoreCheck` is false, but `needsNullValueCheck` is true, the helper call node in `callTree` is never removed.

Reply:
By the way, I know that today `needsStoreCheck` is never false when `needsNullValueCheck` is true, but it will be possible with changes that I'm working on.

Reply:
>I know that today needsStoreCheck is never false when needsNullValueCheck is true, but it will be possible with changes that I'm working on.

Right, I assumed whenever `needsNullValueCheck` is set as true, `needsStoreCheck` is set as true as well. Does the following make more sense? 

```
if (needsStoreCheck)
   {
   ...
   callTree->setNode(storeCheckNode);
   optimizer()->setAliasSetsAreValid(false);
   }
else
   {
   callTree->setNode(TR::Node::create(TR::treetop, 1, elementStoreNode));
   } 

if (needsNullValueCheck)
   {
   ...
   callTree->insertBefore(TR::TreeTop::create(comp(), TR::Node::create(TR::treetop, 1,  nullCheckNode)));
   optimizer()->setAliasSetsAreValid(false);
   }
```

Reply:
Yes, I think that will work.  Thanks, Annabelle @a7ehuo!

Reply:
Thanks for the clarification @hzongaro! [The above change](https://github.com/eclipse-openj9/openj9/pull/14098#discussion_r791161292) is addressed in 5c12ea4